src/core/sprinkler.rs

```diff
@@ -0,0 +1,133 @@
+
+use futures::future::*;
+use futures::stream::*;
+
+
+
+
+
+#[derive(Clone)]
+pub struct Cluster {
+    pub vdisks: Vec<VDisk>
+
+}
+
+#[derive(Clone)]
+pub struct Sprinkler {
+    cluster: Cluster,
+    quorum: u8,
+    link_manager: Arc<LinkManager>
+}
+
+pub struct SprinklerError {
+    total_ops: u16,
+    ok_ops: u16,
+    quorum: u8
+
+}
+
+pub struct SprinklerResult {
+    total_ops: u16,
+    ok_ops: u16,
+    quorum: u8
+}
+
+impl Sprinkler {
+    pub fn new() -> Sprinkler {
+        let ex_cluster = Cluster {
+                vdisks: vec![VDisk {
+                    id: 0,
+                    replicas: vec![NodeDisk {
+                        node: Node {
+                            host: "127.0.0.1".to_string(),
+                            port: 20000
+                        },
+                        path: "/tmp/disk1".to_string()
+                    },
+                    NodeDisk {
+                        node: Node {
+                            host: "127.0.0.1".to_string(),
+                            port: 20002
+                        },
+                        path: "/tmp/disk2".to_string()
+                    }]
+                }]
+            };
+        let nodes: Vec<_> = ex_cluster.vdisks.iter()
+                                    .flat_map(|vdisk| vdisk.replicas.iter().map(|nd| nd.node.clone())  )
+                                    .collect();
+        Sprinkler {
+            quorum: 1,
+            cluster: ex_cluster,
+            link_manager: Arc::new(LinkManager::new(nodes, Duration::from_millis(3000)))
+        }
+    }
+
+    pub fn get_periodic_tasks(&self, ex: tokio::runtime::TaskExecutor) -> Box<impl Future<Item=(), Error=()>> {
+        self.link_manager.get_checker_future(ex)
+    }
+
+    pub fn put_clustered(&self, key: BobKey, data: BobData) 
+            -> impl Future<Item = SprinklerResult, Error = SprinklerError> + 'static + Send {
+        println!("PUT[{:?}]: Data size: {:?}", key, data.data.len());
+        
+        let target_vdisks: Vec<VDisk> = self.cluster.vdisks.iter()
+            .filter(|disk| disk.id == 0)
+            .cloned()
+            .collect();
+
+        
+        println!("PUT[{:?}]: Will use this vdisks {:?}", key, target_vdisks);
+
+        let mut target_nodes: Vec<Node> = target_vdisks.iter()
+            .flat_map(|node_disk| node_disk.replicas.iter()
+                                                    .map(|nd| nd.node.clone())
+                                                    .collect::<Vec<Node>>())
+            .collect();
+        target_nodes.dedup();
+
+        println!("PUT[{:?}]: Nodes for fan out: {:?}", key, target_nodes);
+
+        // incupsulate vdisk in transaction
+        let mut conn_to_send: Vec<_> = target_nodes.iter()
+                                        .map(|n| self.link_manager.clone().get_link(n))
+                                        .collect();
+
+        let reqs: Vec<_> = conn_to_send.iter_mut().map(move |nl| {
```

Comment:
This is the 4th creation of the `Vec<>` in order. Probably they can be combined somehow without materializing into `Vec<>`

Reply:
I think it's possible, but to log what vdisk will be used i need to materialize them. trade off)